src/operations/client_bulk_write/command_builder.ts

```diff
@@ -76,27 +85,59 @@ export class ClientBulkWriteCommandBuilder {
     return this.currentModelIndex < this.models.length;
   }
 
+  /**
+   * When we need to retry a command we need to set the current
+   * model index back to its previous value.
+   */
+  resetBatch(): boolean {
+    this.currentModelIndex = this.previousModelIndex;
+    return true;
+  }
+
   /**
    * Build a single batch of a client bulk write command.
    * @param maxMessageSizeBytes - The max message size in bytes.
    * @param maxWriteBatchSize - The max write batch size.
    * @returns The client bulk write command.
    */
-  buildBatch(maxMessageSizeBytes: number, maxWriteBatchSize: number): ClientBulkWriteCommand {
+  buildBatch(
+    maxMessageSizeBytes: number,
+    maxWriteBatchSize: number,
+    maxBsonObjectSize: number
+  ): ClientBulkWriteCommand {
+    // We start by assuming the batch has no multi-updates, so it is retryable
+    // until we find them.
+    this.isBatchRetryable = true;
     let commandLength = 0;
     let currentNamespaceIndex = 0;
     const command: ClientBulkWriteCommand = this.baseCommand();
     const namespaces = new Map<string, number>();
+    // In the case of retries we need to mark where we started this batch.
+    this.previousModelIndex = this.currentModelIndex;
 
     while (this.currentModelIndex < this.models.length) {
       const model = this.models[this.currentModelIndex];
       const ns = model.namespace;
       const nsIndex = namespaces.get(ns);
 
+      // Multi updates are not retryable.
+      if (model.name === 'deleteMany' || model.name === 'updateMany') {
+        this.isBatchRetryable = false;
+      }
+
       if (nsIndex != null) {
         // Build the operation and serialize it to get the bytes buffer.
         const operation = buildOperation(model, nsIndex, this.pkFactory);
-        const operationBuffer = BSON.serialize(operation);
+        let operationBuffer;
+        try {
+          operationBuffer = BSON.serialize(operation);
+        } catch (error) {
+          throw new MongoInvalidArgumentError(
+            `Could not serialize operation to BSON: ${error.message}.`
+          );
+        }
```

Comment:
```suggestion
        try {
          operationBuffer = BSON.serialize(operation);
        } catch (cause) {
          throw new MongoInvalidArgumentError(
            `Could not serialize operation to BSON`, { cause }
          );
        }
```

Same for the other try catching around BSON, I think keeping the stack trace in the error is worth having because this error should be so rare such that we want the most information when it occurs.

Reply:
`MongoInvalidArgumentError`, which we had decided last week was the appropriate error for this case, only takes a string message in its constructor. This is why I was including the original message in it and not providing a `cause` option. Is your suggestion to create a new error that provide a cause option or modify the existing error to accept it?

Reply:
I would modify the existing one, we've added cause to errors as we've encountered the need, its parent MongoAPIError already supports it so only one level of changes is needed here.

Reply:
Updated InvalidArgumentError to accept a cause and changed the 2 usages here.